File: sdk/python/src/remount/agent.py
"""JSON HTTP clients for Remount Agents and durable approvals."""

from __future__ import annotations

import asyncio
import json as jsonlib
import secrets
from typing import Any, AsyncIterator, Awaitable, Callable, NotRequired, TypedDict
from urllib.parse import quote, urlencode, urlsplit, urlunsplit

import httpx
from websockets.asyncio.client import connect as websocket_connect

from .client import ConnectionClosed, ProtocolError
from .types import Agent, AgentCreateReq, AgentMessageRes, Approval

MAX_HTTP_RESPONSE_BYTES = 16 << 20


class ApprovalDecisionInput(TypedDict):
    """The HTTP decision body; the approval id is carried by the URL."""

    option: NotRequired[str]
    denied: NotRequired[bool]
    content: NotRequired[Any]
    remember: NotRequired[str]


def _key() -> str:
    return "idem_" + secrets.token_hex(16)


class Terminal:
    """One authenticated Agent terminal attachment with an explicit replay cursor."""

    def __init__(self, socket: Any, *, timeout: float, max_message_bytes: int):
        self._socket = socket
        self.timeout = timeout
        self.max_message_bytes = max_message_bytes
        self.session = ""
        self.next_seq = 0
        self._opened = False
        self._closed = False
        self._socket_closed = False

    async def __aenter__(self) -> Terminal:
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.close()

    def __aiter__(self) -> AsyncIterator[dict[str, Any]]:
        return self

    async def __anext__(self) -> dict[str, Any]:
        if self._closed:
            raise StopAsyncIteration
        try:
            message = await asyncio.wait_for(self._socket.recv(), timeout=self.timeout)
        except StopAsyncIteration:
            self._closed = True
            raise
        except TimeoutError:
            await self.close()
            raise ProtocolError("timeout", "terminal receive deadline exceeded") from None
        except asyncio.CancelledError:
            raise
        except Exception:
            self._closed = True
            raise ConnectionClosed("terminal disconnected; reconnect from next_seq") from None
        if isinstance(message, str):
            if len(message.encode()) > self.max_message_bytes:
                await self.close()
                raise ProtocolError("resource_exhausted", "terminal message exceeds configured limit")
            try:
                event = jsonlib.loads(message)
            except ValueError:
                await self.close()
                raise ProtocolError("bad_request", "terminal sent invalid JSON") from None
            if not isinstance(event, dict) or not isinstance(event.get("type"), str):
                await self.close()
                raise ProtocolError("bad_request", "terminal sent invalid control event")
            if not self._opened and event["type"] != "open":
                await self.close()
                raise ProtocolError("bad_request", "terminal first event was not open")
            if event["type"] == "open":
                if self._opened:
                    await self.close()
                    raise ProtocolError("bad_request", "terminal sent a duplicate open event")
                self._opened = True
                self.session = str(event.get("session", ""))
                self.next_seq = int(event.get("next", 0))
            elif event["type"] == "gap":
                self.next_seq = max(self.next_seq, int(event.get("to", 0)) + 1)
            elif event["type"] == "exit":
                self.next_seq += 1
                self._closed = True
            return event
        if not isinstance(message, (bytes, bytearray, memoryview)):
            await self.close()
            raise ProtocolError("bad_request", "terminal sent an unsupported message")
        data = bytes(message)
        if len(data) > self.max_message_bytes:
            await self.close()
            raise ProtocolError("resource_exhausted", "terminal message exceeds configured limit")
        if not self._opened:
            await self.close()
            raise ProtocolError("bad_request", "terminal output preceded open")
        self.next_seq += 1
        return {"type": "data", "data": data}

    async def input(self, data: bytes) -> None:
        await self._socket.send(data)

    async def resize(self, rows: int, cols: int) -> None:
        await self._control({"type": "resize", "rows": rows, "cols": cols})

    async def signal(self, signal: str) -> None:
        await self._control({"type": "signal", "signal": signal})

    async def eof(self) -> None:
        await self._control({"type": "eof"})

    async def _control(self, event: dict[str, Any]) -> None:
        await self._socket.send(jsonlib.dumps(event, separators=(",", ":")))

    async def close(self) -> None:
        self._closed = True
        if not self._socket_closed:
            self._socket_closed = True
            await self._socket.close()


class AgentClient:
    """Async HTTP client for agent, transcript, and approval resources."""

    def __init__(self, base_url: str, token: str, *, client: httpx.AsyncClient | None = None, max_response_bytes: int = MAX_HTTP_RESPONSE_BYTES, retries: int = 3, terminal_connector: Callable[..., Awaitable[Any]] | None = None):
        if retries < 1 or retries > 10:
            raise ValueError("retries must be between 1 and 10")
        if max_response_bytes < 1:
            raise ValueError("max_response_bytes must be positive")
        self.base_url = base_url.rstrip("/")
        self.token = token
        self._client = client
        self._owns_client = client is None
        self.max_response_bytes = max_response_bytes
        self.retries = retries
        self._terminal_connector = terminal_connector or websocket_connect

    async def __aenter__(self) -> AgentClient:
        await self._http()
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.close()

    async def close(self) -> None:
        """Close the internally owned HTTP connection pool."""
        if self._client is not None and self._owns_client:
            await self._client.aclose()
            self._client = None

    async def _http(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=30, follow_redirects=False)
        return self._client

    async def _request(self, method: str, path: str, *, json: Any = None, idempotency_key: str | None = None) -> Any:
        headers = {"Authorization": f"Bearer {self.token}"}
        if method != "GET":
            headers["Idempotency-Key"] = idempotency_key or _key()
        request: dict[str, Any] = {"headers": headers}
        if json is not None:
            request["json"] = json
        for attempt in range(self.retries):
            try:
                async with (await self._http()).stream(method, self.base_url + path, **request) as response:
                    chunks: list[bytes] = []
                    size = 0
                    async for chunk in response.aiter_bytes():
                        size += len(chunk)
                        if size > self.max_response_bytes:
                            raise ProtocolError("resource_exhausted", "HTTP response exceeds configured limit")
                        chunks.append(chunk)
                    raw = b"".join(chunks)
                    status = response.status_code
                    success = response.is_success
                break
            except httpx.TransportError:
                if attempt + 1 == self.retries:
                    raise ProtocolError("unreachable", "HTTP request failed") from None
                await asyncio.sleep(0.1 * (attempt + 1))
        if not success:
            try:
                body = jsonlib.loads(raw)
            except (ValueError, UnicodeDecodeError):
                body = {}
            detail = body.get("error", body) if isinstance(body, dict) else {}
            if not isinstance(detail, dict):
                detail = {}
            message = str(detail.get("message", f"HTTP {status}"))
            if self.token:
                message = message.replace(self.token, "[redacted]")
            raise ProtocolError(str(detail.get("code", "internal")), message)
        if status == 204:
            return None
        try:
            return jsonlib.loads(raw)
        except (ValueError, UnicodeDecodeError) as error:
            raise ProtocolError("internal", "server returned invalid JSON") from error

    async def create_agent(self, request: AgentCreateReq, *, idempotency_key: str | None = None) -> Agent:
        return await self._request("POST", "/v1/agents", json=request, idempotency_key=idempotency_key)

    async def get_agent(self, agent_id: str) -> Agent:
        return await self._request("GET", "/v1/agents/" + quote(agent_id, safe=""))

    async def list_agents(self, *, status: str = "", workspace: str = "", parent: str = "") -> list[Agent]:
        params = httpx.QueryParams({k: v for k, v in {"status": status, "ws": workspace, "parent": parent}.items() if v})
        result = await self._request("GET", "/v1/agents" + ("?" + str(params) if params else ""))
        return result["agents"]

    async def message_agent(self, agent_id: str, text: str, *, kind: str = "", idempotency_key: str | None = None) -> AgentMessageRes:
        return await self._request("POST", f"/v1/agents/{quote(agent_id, safe='')}/messages", json={"text": text, "kind": kind}, idempotency_key=idempotency_key)

    async def fork_agent(self, agent_id: str, request: dict[str, Any] | None = None, *, idempotency_key: str | None = None) -> Agent:
        return await self._request("POST", f"/v1/agents/{quote(agent_id, safe='')}/fork", json=request or {}, idempotency_key=idempotency_key)

    async def agent_action(self, agent_id: str, action: str, *, by: str = "", idempotency_key: str | None = None) -> Agent | None:
        if action not in {"cancel", "sleep", "wake", "destroy"}:
            raise ValueError("unsupported agent action")
        return await self._request("POST", f"/v1/agents/{quote(agent_id, safe='')}/{action}", json={"by": by} if action == "wake" else None, idempotency_key=idempotency_key)

    async def transcript(self, agent_id: str, *, from_index: int = 0, limit: int = 0) -> dict[str, Any]:
        params = httpx.QueryParams({"from": from_index, "limit": limit})
        return await self._request("GET", f"/v1/agents/{quote(agent_id, safe='')}/transcript?{params}")

    async def stream_transcript(self, agent_id: str, *, from_index: int = 0, poll_interval: float = 0.2) -> AsyncIterator[dict[str, Any]]:
        cursor = from_index
        while True:
            try:
                page = await self.transcript(agent_id, from_index=cursor)
            except ProtocolError as error:
                if error.code != "unreachable":
                    raise
                await asyncio.sleep(poll_interval)
                continue
            if page.get("gap"):
                gap = page["gap"]
                raise ProtocolError("evicted", f"transcript records [{gap['from']}, {gap['to']}) were evicted", int(gap["to"]))
            for record in page.get("records", []):
                yield record
            next_cursor = int(page.get("next", cursor))
            if next_cursor < cursor:
                raise ProtocolError("internal", "transcript cursor moved backwards")
            cursor = next_cursor
            if page.get("done"):
                return
            await asyncio.sleep(poll_interval)

    async def list_approvals(self, agent_id: str, *, status: str = "") -> list[Approval]:
        suffix = "?" + str(httpx.QueryParams({"status": status})) if status else ""
        result = await self._request("GET", f"/v1/agents/{quote(agent_id, safe='')}/approvals{suffix}")
        return result["approvals"]

    async def get_approval(self, approval_id: str) -> Approval:
        return await self._request("GET", "/v1/approvals/" + quote(approval_id, safe=""))

    async def decide_approval(self, approval_id: str, decision: ApprovalDecisionInput, *, idempotency_key: str | None = None) -> Approval:
        return await self._request("POST", "/v1/approvals/" + quote(approval_id, safe=""), json=decision, idempotency_key=idempotency_key)

    async def approve(self, approval_id: str, option: str = "allow_once", *, idempotency_key: str | None = None) -> Approval:
        """Choose one offered approval option."""
        return await self.decide_approval(approval_id, {"option": option}, idempotency_key=idempotency_key)

    async def deny(self, approval_id: str, *, idempotency_key: str | None = None) -> Approval:
        """Deny a pending approval."""
        return await self.decide_approval(approval_id, {"denied": True}, idempotency_key=idempotency_key)

    async def wait_for_agent(self, agent_id: str, statuses: set[str], *, timeout: float = 120, poll_interval: float = 0.2) -> Agent:
        """Poll the read-only Agent resource until it reaches one of statuses."""
        try:
            async with asyncio.timeout(timeout):
                while True:
                    agent = await self.get_agent(agent_id)
                    if agent.get("status") in statuses:
                        return agent
                    await asyncio.sleep(poll_interval)
        except TimeoutError:
            raise ProtocolError("timeout", "agent status wait deadline exceeded") from None

    async def wait_for_approval(self, agent_id: str, *, timeout: float = 120, poll_interval: float = 0.2) -> Approval:
        """Poll without waking the Agent until its first pending approval exists."""
        try:
            async with asyncio.timeout(timeout):
                while True:
                    approvals = await self.list_approvals(agent_id)
                    if approvals:
                        return approvals[0]
                    await asyncio.sleep(poll_interval)
        except TimeoutError:
            raise ProtocolError("timeout", "approval wait deadline exceeded") from None

    async def diff(self, agent_id: str, *, wake: bool = False) -> dict[str, Any]:
        """Read the bounded unified diff, optionally waking a sleeping Agent."""
        suffix = "?wake=true" if wake else ""
        return await self._request("GET", f"/v1/agents/{quote(agent_id, safe='')}/diff{suffix}")

    async def connect_terminal(
        self,
        agent_id: str,
        *,
        session: str = "",
        from_seq: int = 0,
        program: list[str] | None = None,
        cwd: str = "",
        rows: int = 24,
        cols: int = 80,
        timeout: float = 30,
    ) -> Terminal:
        """Attach a terminal; reconnect with returned session and next_seq."""
        if timeout <= 0:
            raise ValueError("timeout must be positive")
        query: list[tuple[str, str]] = []
        if session:
            query.extend((("session", session), ("from", str(from_seq))))
        else:
            query.extend(("program", value) for value in (program or ["/bin/sh"]))
            query.extend((("cwd", cwd), ("rows", str(rows)), ("cols", str(cols))))
        base = urlsplit(self.base_url)
        if base.scheme not in {"http", "https", "ws", "wss"} or not base.netloc:
            raise ValueError("Remount URL must use http, https, ws, or wss")
        path = base.path.rstrip("/") + f"/v1/agents/{quote(agent_id, safe='')}/terminal"
        scheme = "wss" if base.scheme in {"https", "wss"} else "ws"
        url = urlunsplit((scheme, base.netloc, path, urlencode(query), ""))
        try:
            socket = await asyncio.wait_for(
                self._terminal_connector(
                    url,
                    additional_headers={"Authorization": f"Bearer {self.token}"},
                    compression=None,
                    max_size=self.max_response_bytes,
                    max_queue=16,
                ),
                timeout=timeout,
            )
        except TimeoutError:
            raise ProtocolError("timeout", "terminal connect deadline exceeded") from None
        except asyncio.CancelledError:
            raise
        except Exception:
            raise ConnectionClosed("terminal connection failed") from None
        return Terminal(socket, timeout=timeout, max_message_bytes=self.max_response_bytes)

File: sdk/typescript/src/agent.ts
import { ConnectionClosed } from "./client.js";
import { ProtocolError } from "./errors.js";
import type { Agent, AgentCreateReq, AgentMessageRes, Approval, ApprovalDecideReq } from "./types.js";

function idem(): string { return `idem_${globalThis.crypto.randomUUID().replaceAll("-", "")}`; }

export interface TerminalWebSocket {
  binaryType?: string;
  send(data: string | Uint8Array): void;
  close(): void;
  addEventListener?(name: string, listener: (event: any) => void): void;
  on?(name: string, listener: (...args: any[]) => void): void;
}

export type TerminalWebSocketFactory = (url: string, bearerProtocol: string) => Promise<TerminalWebSocket>;

function bearerProtocol(token: string): string {
  let binary = "";
  for (const value of new TextEncoder().encode(token)) binary += String.fromCharCode(value);
  return `remount.bearer.${btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "")}`;
}

async function defaultTerminalFactory(url: string, protocol: string): Promise<TerminalWebSocket> {
  const Constructor: any = globalThis.WebSocket ?? (await import("ws")).default;
  return await new Promise((resolve, reject) => {
    const socket: TerminalWebSocket = new Constructor(url, protocol);
    socket.binaryType = "arraybuffer";
    const open = () => resolve(socket);
    const error = () => reject(new ConnectionClosed("terminal WebSocket handshake failed"));
    if (socket.addEventListener) {
      socket.addEventListener("open", open); socket.addEventListener("error", error);
    } else {
      socket.on!("open", open); socket.on!("error", error);
    }
  });
}

function terminalBytes(value: any): Uint8Array {
  if (value instanceof Uint8Array) return value;
  if (value instanceof ArrayBuffer) return new Uint8Array(value);
  if (ArrayBuffer.isView(value)) return new Uint8Array(value.buffer, value.byteOffset, value.byteLength);
  throw new ProtocolError("bad_request", "terminal sent an unsupported message");
}

export interface TerminalEvent { type: string; data?: Uint8Array; [key: string]: unknown }

/** The HTTP decision body; the approval id and idempotency key use headers/path. */
export type ApprovalDecisionInput = Omit<ApprovalDecideReq, "id" | "idem">;

/** One authenticated Agent terminal attachment with an explicit replay cursor. */
export class Terminal implements AsyncIterable<TerminalEvent> {
  session = "";
  nextSeq = 0;
  private opened = false;
  private ended = false;
  private error?: Error;
  private queue: TerminalEvent[] = [];
  private queueBytes = 0;
  private waiters: Array<{ resolve: (value: IteratorResult<TerminalEvent>) => void; reject: (error: Error) => void }> = [];

  constructor(private socket: TerminalWebSocket, private maxMessageBytes: number) {
    if (socket.addEventListener) {
      socket.addEventListener("message", (event) => this.receive(event.data));
      socket.addEventListener("close", () => this.disconnected());
      socket.addEventListener("error", () => this.fail(new ConnectionClosed("terminal connection failed")));
    } else {
      socket.on!("message", (data, isBinary) => this.receive(isBinary ? data : new TextDecoder().decode(terminalBytes(data))));
      socket.on!("close", () => this.disconnected());
      socket.on!("error", () => this.fail(new ConnectionClosed("terminal connection failed")));
    }
  }

  [Symbol.asyncIterator](): AsyncIterator<TerminalEvent> { return { next: () => this.next() }; }

  private next(): Promise<IteratorResult<TerminalEvent>> {
    if (this.queue.length) {
      const event = this.queue.shift()!;
      this.queueBytes -= event.data?.byteLength ?? 0;
      return Promise.resolve({ value: event, done: false });
    }
    if (this.ended) return this.error ? Promise.reject(this.error) : Promise.resolve({ value: undefined, done: true });
    if (this.waiters.length >= 1024) {
      const error = new ProtocolError("resource_exhausted", "too many pending terminal consumers");
      this.fail(error); return Promise.reject(error);
    }
    return new Promise((resolve, reject) => this.waiters.push({ resolve, reject }));
  }

  private receive(value: any): void {
    if (this.ended) return;
    try {
      if (typeof value === "string") {
        if (new TextEncoder().encode(value).byteLength > this.maxMessageBytes) throw new ProtocolError("resource_exhausted", "terminal message exceeds configured limit");
        const event = JSON.parse(value) as TerminalEvent;
        if (!event || typeof event.type !== "string") throw new ProtocolError("bad_request", "terminal sent invalid control event");
        if (!this.opened && event.type !== "open") throw new ProtocolError("bad_request", "terminal first event was not open");
        if (event.type === "open") {
          if (this.opened) throw new ProtocolError("bad_request", "terminal sent a duplicate open event");
          this.opened = true; this.session = String(event.session ?? ""); this.nextSeq = Number(event.next ?? 0);
        } else if (event.type === "gap") this.nextSeq = Math.max(this.nextSeq, Number(event.to ?? 0) + 1);
        else if (event.type === "exit") { this.nextSeq++; this.ended = true; }
        this.push(event);
        if (this.ended) this.finishWaiters();
        return;
      }
      const data = terminalBytes(value);
      if (data.byteLength > this.maxMessageBytes) throw new ProtocolError("resource_exhausted", "terminal message exceeds configured limit");
      if (!this.opened) throw new ProtocolError("bad_request", "terminal output preceded open");
      this.nextSeq++;
      this.push({ type: "data", data });
    } catch (error) {
      this.fail(error instanceof Error ? error : new ProtocolError("bad_request", "invalid terminal message"));
      this.socket.close();
    }
  }

  private push(event: TerminalEvent): void {
    const waiter = this.waiters.shift();
    if (waiter) { waiter.resolve({ value: event, done: false }); return; }
    const size = event.data?.byteLength ?? 0;
    if (this.queue.length >= 1024 || this.queueBytes + size > this.maxMessageBytes) {
      this.fail(new ProtocolError("resource_exhausted", "terminal delivery queue is full")); this.socket.close(); return;
    }
    this.queue.push(event); this.queueBytes += size;
  }

  private disconnected(): void {
    if (!this.ended) this.fail(new ConnectionClosed("terminal disconnected; reconnect from nextSeq"));
  }

  private fail(error: Error): void {
    if (this.ended) return;
    this.error = error; this.ended = true; this.finishWaiters();
  }

  private finishWaiters(): void {
    for (const waiter of this.waiters.splice(0)) {
      if (this.error) waiter.reject(this.error);
      else waiter.resolve({ value: undefined, done: true });
    }
  }

  input(data: Uint8Array): void { this.socket.send(data); }
  resize(rows: number, cols: number): void { this.control({ type: "resize", rows, cols }); }
  signal(signal: string): void { this.control({ type: "signal", signal }); }
  eof(): void { this.control({ type: "eof" }); }
  private control(event: Record<string, unknown>): void { this.socket.send(JSON.stringify(event)); }
  close(): void { this.ended = true; this.finishWaiters(); this.socket.close(); }
}

export interface AgentClientOptions { fetch?: typeof fetch; requestTimeoutMilliseconds?: number; maxResponseBytes?: number; retries?: number; terminalWebSocketFactory?: TerminalWebSocketFactory }

export class AgentClient {
  private fetcher: typeof fetch;
  private requestTimeoutMilliseconds: number;
  private maxResponseBytes: number;
  private retries: number;
  private terminalFactory: TerminalWebSocketFactory;
  readonly baseURL: string;
  constructor(baseURL: string, readonly token: string, options: AgentClientOptions = {}) {
    this.baseURL = baseURL.replace(/\/$/, "");
    this.fetcher = options.fetch ?? globalThis.fetch.bind(globalThis);
    this.requestTimeoutMilliseconds = options.requestTimeoutMilliseconds ?? 30_000;
    this.maxResponseBytes = options.maxResponseBytes ?? 16 << 20;
    this.retries = options.retries ?? 3;
    if (this.retries < 1 || this.retries > 10) throw new Error("retries must be between 1 and 10");
    if (this.maxResponseBytes < 1) throw new Error("maxResponseBytes must be positive");
    if (this.requestTimeoutMilliseconds <= 0) throw new Error("requestTimeoutMilliseconds must be positive");
    this.terminalFactory = options.terminalWebSocketFactory ?? defaultTerminalFactory;
  }

  private async request(method: string, path: string, body?: unknown, idempotencyKey?: string): Promise<any> {
    const headers: Record<string, string> = { Authorization: `Bearer ${this.token}` };
    if (method !== "GET") headers["Idempotency-Key"] = idempotencyKey ?? idem();
    if (body !== undefined) headers["Content-Type"] = "application/json";
    for (let attempt = 0; attempt < this.retries; attempt++) {
      try {
        const signal = this.requestTimeoutMilliseconds > 0 ? AbortSignal.timeout(this.requestTimeoutMilliseconds) : undefined;
        const response = await this.fetcher(this.baseURL + path, { method, headers, body: body === undefined ? undefined : JSON.stringify(body), redirect: "error", signal });
        if (response.status === 204 && response.ok) return undefined;
        const raw = await readBounded(response, this.maxResponseBytes);
        let payload: any = {};
        try { payload = JSON.parse(new TextDecoder().decode(raw)); } catch {
          if (response.ok) throw new ProtocolError("internal", "server returned invalid JSON");
        }
        if (!response.ok) {
          const detail = payload.error && typeof payload.error === "object" ? payload.error : payload;
          const message = String(detail.message ?? `HTTP ${response.status}`).split(this.token).join(this.token ? "[redacted]" : "");
          throw new ProtocolError(String(detail.code ?? "internal"), message);
        }
        return payload;
      } catch (error) {
        if (error instanceof ProtocolError) throw error;
        if (attempt + 1 === this.retries) throw new ProtocolError("unreachable", "HTTP request failed");
        await new Promise((resolve) => setTimeout(resolve, 100 * (attempt + 1)));
      }
    }
    throw new ProtocolError("unreachable", "HTTP request failed");
  }

  createAgent(request: AgentCreateReq, idempotencyKey?: string): Promise<Agent> { return this.request("POST", "/v1/agents", request, idempotencyKey); }
  getAgent(id: string): Promise<Agent> { return this.request("GET", `/v1/agents/${encodeURIComponent(id)}`); }
  async listAgents(filters: { status?: string; workspace?: string; parent?: string } = {}): Promise<Agent[]> {
    const query = new URLSearchParams();
    if (filters.status) query.set("status", filters.status);
    if (filters.workspace) query.set("ws", filters.workspace);
    if (filters.parent) query.set("parent", filters.parent);
    const result = await this.request("GET", `/v1/agents${query.size ? `?${query}` : ""}`);
    return result.agents;
  }
  messageAgent(id: string, text: string, options: { kind?: string; idempotencyKey?: string } = {}): Promise<AgentMessageRes> { return this.request("POST", `/v1/agents/${encodeURIComponent(id)}/messages`, { text, kind: options.kind ?? "" }, options.idempotencyKey); }
  forkAgent(id: string, request: Record<string, unknown> = {}, idempotencyKey?: string): Promise<Agent> { return this.request("POST", `/v1/agents/${encodeURIComponent(id)}/fork`, request, idempotencyKey); }
  agentAction(id: string, action: "cancel" | "sleep" | "wake" | "destroy", options: { by?: string; idempotencyKey?: string } = {}): Promise<Agent | undefined> { return this.request("POST", `/v1/agents/${encodeURIComponent(id)}/${action}`, action === "wake" ? { by: options.by ?? "" } : undefined, options.idempotencyKey); }

  transcript(id: string, from = 0, limit = 0): Promise<any> {
    const query = new URLSearchParams({ from: String(from), limit: String(limit) });
    return this.request("GET", `/v1/agents/${encodeURIComponent(id)}/transcript?${query}`);
  }

  async *streamTranscript(id: string, options: { from?: number; pollMilliseconds?: number } = {}): AsyncGenerator<any> {
    let cursor = options.from ?? 0;
    for (;;) {
      let page: any;
      try { page = await this.transcript(id, cursor); }
      catch (error) {
        if (!(error instanceof ProtocolError) || error.code !== "unreachable") throw error;
        await new Promise((resolve) => setTimeout(resolve, options.pollMilliseconds ?? 200));
        continue;
      }
      if (page.gap) throw new ProtocolError("evicted", `transcript records [${page.gap.from}, ${page.gap.to}) were evicted`, Number(page.gap.to));
      for (const record of page.records ?? []) yield record;
      const next = Number(page.next ?? cursor);
      if (next < cursor) throw new ProtocolError("internal", "transcript cursor moved backwards");
      cursor = next;
      if (page.done) return;
      await new Promise((resolve) => setTimeout(resolve, options.pollMilliseconds ?? 200));
    }
  }

  async listApprovals(agentID: string, status = ""): Promise<Approval[]> {
    const query = status ? `?${new URLSearchParams({ status })}` : "";
    const result = await this.request("GET", `/v1/agents/${encodeURIComponent(agentID)}/approvals${query}`);
    return result.approvals;
  }
  getApproval(id: string): Promise<Approval> { return this.request("GET", `/v1/approvals/${encodeURIComponent(id)}`); }
  decideApproval(id: string, decision: ApprovalDecisionInput, idempotencyKey?: string): Promise<Approval> { return this.request("POST", `/v1/approvals/${encodeURIComponent(id)}`, decision, idempotencyKey); }
  approve(id: string, option = "allow_once", idempotencyKey?: string): Promise<Approval> { return this.decideApproval(id, { option }, idempotencyKey); }
  deny(id: string, idempotencyKey?: string): Promise<Approval> { return this.decideApproval(id, { denied: true }, idempotencyKey); }

  async waitForAgent(id: string, statuses: ReadonlySet<string>, options: { timeoutMilliseconds?: number; pollMilliseconds?: number } = {}): Promise<Agent> {
    const deadline = Date.now() + (options.timeoutMilliseconds ?? 120_000);
    for (;;) {
      const agent = await this.getAgent(id);
      if (statuses.has(agent.status)) return agent;
      if (Date.now() >= deadline) throw new ProtocolError("timeout", "agent status wait deadline exceeded");
      await new Promise((resolve) => setTimeout(resolve, options.pollMilliseconds ?? 200));
    }
  }

  async waitForApproval(agentID: string, options: { timeoutMilliseconds?: number; pollMilliseconds?: number } = {}): Promise<Approval> {
    const deadline = Date.now() + (options.timeoutMilliseconds ?? 120_000);
    for (;;) {
      const approvals = await this.listApprovals(agentID);
      if (approvals.length) return approvals[0];
      if (Date.now() >= deadline) throw new ProtocolError("timeout", "approval wait deadline exceeded");
      await new Promise((resolve) => setTimeout(resolve, options.pollMilliseconds ?? 200));
    }
  }

  diff(id: string, wake = false): Promise<{ status: string; diff: string; truncated: boolean }> {
    return this.request("GET", `/v1/agents/${encodeURIComponent(id)}/diff${wake ? "?wake=true" : ""}`);
  }

  async connectTerminal(id: string, options: { session?: string; from?: number; program?: string[]; cwd?: string; rows?: number; cols?: number; timeoutMilliseconds?: number } = {}): Promise<Terminal> {
    const url = new URL(this.baseURL);
    if (!["http:", "https:", "ws:", "wss:"].includes(url.protocol)) throw new Error("Remount URL must use http, https, ws, or wss");
    url.protocol = url.protocol === "https:" || url.protocol === "wss:" ? "wss:" : "ws:";
    url.pathname = `${url.pathname.replace(/\/$/, "")}/v1/agents/${encodeURIComponent(id)}/terminal`;
    url.search = "";
    if (options.session) {
      url.searchParams.set("session", options.session); url.searchParams.set("from", String(options.from ?? 0));
    } else {
      for (const command of options.program ?? ["/bin/sh"]) url.searchParams.append("program", command);
      url.searchParams.set("cwd", options.cwd ?? "");
      url.searchParams.set("rows", String(options.rows ?? 24)); url.searchParams.set("cols", String(options.cols ?? 80));
    }
    const timeout = options.timeoutMilliseconds ?? this.requestTimeoutMilliseconds;
    if (timeout <= 0) throw new Error("timeoutMilliseconds must be positive");
    const pending = this.terminalFactory(url.toString(), bearerProtocol(this.token));
    let timedOut = false; let handle: ReturnType<typeof setTimeout>;
    const deadline = new Promise<never>((_, reject) => { handle = setTimeout(() => { timedOut = true; reject(new ProtocolError("timeout", "terminal connect deadline exceeded")); }, timeout); });
    pending.then((late) => { if (timedOut) late.close(); }, () => undefined);
    let socket: TerminalWebSocket;
    try { socket = await Promise.race([pending, deadline]); }
    catch (error) {
      if (error instanceof ProtocolError) throw error;
      throw new ConnectionClosed("terminal connection failed");
    } finally { clearTimeout(handle!); }
    return new Terminal(socket, this.maxResponseBytes);
  }
}

async function readBounded(response: Response, maxBytes: number): Promise<Uint8Array> {
  if (!response.body) return new Uint8Array();
  const reader = response.body.getReader(); const chunks: Uint8Array[] = []; let size = 0;
  for (;;) {
    const { value, done } = await reader.read(); if (done) break;
    size += value.byteLength;
    if (size > maxBytes) { await reader.cancel(); throw new ProtocolError("resource_exhausted", "HTTP response exceeds configured limit"); }
    chunks.push(value);
  }
  const result = new Uint8Array(size); let offset = 0;
  for (const chunk of chunks) { result.set(chunk, offset); offset += chunk.byteLength; }
  return result;
}
